src/internal_telemetry/allocations/allocator/stack.rs

```diff
@@ -0,0 +1,53 @@
+use super::token::AllocationGroupId;
+
+/// An allocation group stack.
+///
+/// As allocation groups are entered and exited, they naturally end up looking a lot like a stack itself: the active
+/// allocation group gets added to the stack when entered, and if another allocation group is entered before the
+/// previous is exited, the newer group is added to the stack above the previous one, and so on and so forth.
+///
+/// This implementation is uses an array to represent the stack to avoid thread local destructor registration issues.
+#[derive(Copy, Clone)]
+pub(crate) struct GroupStack<const N: usize> {
+    slots: [AllocationGroupId; N],
+    current_top: usize,
+}
+
+impl<const N: usize> GroupStack<N> {
+    /// Creates an empty [`GroupStack`].
+    pub const fn new() -> Self {
+        Self {
+            current_top: 0,
+            slots: [AllocationGroupId::ROOT; N],
+        }
+    }
+
+    /// Gets the currently active allocation group.
+    ///
+    /// If the stack is empty, then the root allocation group is the defacto active allocation group, and is returned as such.
+    pub const fn current(&self) -> AllocationGroupId {
+        if self.current_top == 0 {
+            AllocationGroupId::ROOT
+        } else {
+            self.slots[self.current_top - 1]
+        }
+    }
+
+    /// Pushes an allocation group on to the stack, marking it as the active allocation group.
+    pub fn push(&mut self, group: AllocationGroupId) {
+        if self.current_top >= self.slots.len() {
+            panic!("tried to push new allocation group to the current stack, but hit the limit of {} entries", N);
+        }
+        self.slots[self.current_top] = group;
+        self.current_top += 1;
+    }
+
+    /// Pops and returns the previous allocation group that was on the stack.
+    pub fn pop(&mut self) -> AllocationGroupId {
+        if self.current_top == 0 {
+            panic!("tried to pop current allocation group from the stack but the stack is empty");
+        }
+        self.current_top -= 1;
```

Comment:
I wonder if `checked_sub` works better here:
```suggestion
        self.current_top = self.current_top.checked_sub(1)
           .expect("tried to pop current allocation group from the stack but the stack is empty");
```